src/tools/rbd_mirror/InstanceWatcher.cc

```diff
@@ -610,13 +600,9 @@ void InstanceWatcher<I>::handle_release_leader() {
 
   std::lock_guard locker{m_lock};
 
-  ceph_assert(m_image_sync_throttler != nullptr);
-
   m_leader_instance_id.clear();
 
-  m_image_sync_throttler->drain(-ESTALE);
-  m_image_sync_throttler->destroy();
-  m_image_sync_throttler = nullptr;
+  m_image_sync_throttler->drain(m_ioctx.get_namespace(), -ESTALE);
```

Comment:
One of the reasons was a case when we shut down a namespace replayer when its namespace is deleted. Then we would want to drain all sync requests for this namespace. Though, thinking about this, the namespace can not be deleted until all images are deleted so, it might be not an issue (if only there is no some race).

But as far as remember, there was another actual reason. Because actually, in my earlier implementation I did it the way you suggested, but observed an assertion violation with a non-empty requests queue on destroy. In that implementation, in `PoolReplayer::handle_pre_release_leader` I called `handle_release_leader` for every namespace and then called drain. And was observing crashes. I don't remember details now but I attributed them to the changed order (drain started to be called later than it had been previously), so I decided to make it be the same as it is in non-namespace implementation, and it fixed the crashes. 

If you still think it is cleaner to call drain only once for all namespaces, I can try it again and will probably be able to resolve that issue (or provide more details about it).

Reply:
If it simplifies a potential race condition I am fine w/ the current implementation, then.